File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py

from itertools import combinations
from collections import deque


def solution(n, wires):
    answer = n

    def bfs(start):
        visited[start] = True
        queue = deque([start])
        cnt = 1
        while queue:
            cur_v = queue.popleft()
            for v in graph[cur_v]:
                if not visited[v]:
                    visited[v] = True
                    queue.append(v)
                    cnt += 1
        return cnt

    for c in combinations(wires, len(wires) - 1):
        graph = [[] for _ in range(n + 1)]
        visited = [False] * (n + 1)
        for first_node, second_node in c:
            graph[first_node].append(second_node)
            graph[second_node].append(first_node)

        first_network = bfs(1)
        second_network = n - first_network
        answer = min(answer, abs(first_network - second_network))

    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    static List<Integer>[] list;
    static boolean[] visited;
    public int solution(int n, int[][] wires) {
        list = new ArrayList[n+1];
        for (int i = 0; i <= n; i ++){
            list[i] = new ArrayList<>();
        }
        for(int[] wire : wires){
            list[wire[0]].add(wire[1]);
            list[wire[1]].add(wire[0]);
        }
        int res = n;
        for(int[] wire : wires){
            int cnt1 = bfs(wire, n);
            int cnt2 = n - cnt1;
            res = Math.min(res, Math.abs(cnt1-cnt2));
        }
        
        return res;
    }
    private int bfs(int[] wire, int n){
        Deque<Integer> dq = new ArrayDeque<>();
        visited = new boolean[n+1];
        int start = wire[0];
        int skip = wire[1];
        visited[start] = true;
        dq.offer(start);
        int cnt = 1;
        while(!dq.isEmpty()){
            int curr = dq.poll();
            for(int next : list[curr]){
                if(next != skip && !visited[next]){
                    dq.offer(next);
                    visited[next] = true;
                    cnt++;
                }
            }
        }
        
        return cnt;
    }
}